### PR TITLE
Retain exception thrown from close() in SequenceFileRecordWriter

### PR DESCRIPTION
**Pull request by sorin-florea**
Changed the try with resource to use com.google.common.io.Closer to avoid self-suppression from close() in SequenceFileRecordWriter 

Fixes https://github.com/prestosql/presto/issues/3326

### CODE REVIEW

**Comment by floreasorin** [comment]
https://docs.oracle.com/javase/specs/jls/se8/html/jls-14.html#jls-14.20.3.1

**Comment by floreasorin** [comment]
> Thank you for your pull request and welcome to our community. We require contributors to sign our [Contributor License Agreement](https://github.com/prestosql/cla/raw/master/Presto%20Foundation%20Individual%20CLA.pdf), and we don't seem to have you on file. In order for us to review and merge your code, please submit the signed CLA to [cla@prestosql.io](mailto:cla@prestosql.io). For more information, see https://github.com/prestosql/cla.

Cla sent, waiting for approval.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/SequenceFileRecordWriter.java
Comment (losipiuk) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/SequenceFileRecordWriter.java:68]: rename to `finalThrowable`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/SequenceFileRecordWriter.java
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/SequenceFileRecordWriter.java:67]: Thanks for implementing try-with-resources manually. I know it's annoying to write. However, I think we can simplify this using `Closer` from Guava, which avoids self suppression
```java
// use Closer to avoid self-suppression
try (Closer closer = Closer.create()) {
    closer.register(writer);

**Review by electrum — DISMISSED** [review]

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/SequenceFileRecordWriter.java
Comment (floreasorin) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/SequenceFileRecordWriter.java:67]: @electrum that's much cleaner, will add that

Comment (losipiuk) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/SequenceFileRecordWriter.java:67]: I think I am missing something:
In what way is
```java
try (Closeable ignored = writer) {
   ...
```
semantically different from
```java
try (Closer closer = Closer.create()) {
   closer.register(writer);
   ...
```

In both cases try-with-resources is in place and if exception is thrown from `try` block the possible exception thrown from `writer.close()` will be suppressed.

Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/SequenceFileRecordWriter.java:67]: Both options presented here (try-with-resources directly, or with closer) are equivalent from "Self-suppression not permitted" perspective (both end up throwing the "Self-suppression not permitted" instead of actual).

```
public static void main(String[] args)
        throws Exception
{
//        try (Gizmo x = new Gizmo()) {
//            x.thr();
//        }

    try (Closer closer = Closer.create()) {
        Gizmo x = closer.register(new Gizmo());
        x.thr();
    }
}

private static class Gizmo
        implements Closeable
{
    private final RuntimeException e = new RuntimeException("gizmo!!!");

    public void thr()
    {
        throw e;
    }

    @Override
    public void close()
    {
        throw e;
    }
}
```

The way out is using `Closer` the Java 7 way

```java
Closer closer = Closer.create();
Gizmo x = closer.register(new Gizmo());
try {
    x.thr();
}
catch (Exception e) {
    throw closer.rethrow(e);
}
finally {
    closer.close();
}
```

while verbose, this is still simpler than implementing try-with-resource manually, with correct suppression

```
Throwable thrown = null;
Gizmo x = new Gizmo();
try {
    x.thr();
}
catch (Throwable e) {
    thrown = e;
}
finally {
    try {
        x.close();
    }
    catch (Throwable closeException) {
        if (thrown == null) {
            thrown = closeException;
        }
        else if (thrown != closeException) {
            thrown.addSuppressed(closeException);
        }
    }
}
if (thrown != null) {
    propagateIfPossible(thrown, IOException.class);
    throw new RuntimeException(thrown);
}
```

**Review by findepi — COMMENTED** [review]
@floreasorin sorry for lack of review.
can you please change it to "using Closer the Java 7 way" i proposed in the comment?

**Comment by colebow** [comment]
:wave: @sorin-florea - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.

**Pull request closed without merging**
